File: roles/components/mysql_replicaset/files/mysql-replicaset.js
/* global dba, mysql, os, print, shell */

shell.options.useWizards = false;

const action = os.getenv("MYSQL_REPLICASET_ACTION") || "status";
const replicaSetName = os.getenv("MYSQL_REPLICASET_NAME");
const instances = JSON.parse(os.getenv("MYSQL_REPLICASET_INSTANCES") || "[]");
const initialPrimary = os.getenv("MYSQL_REPLICASET_INITIAL_PRIMARY");
const adminUser = os.getenv("MYSQL_REPLICASET_ADMIN_USER");
const adminPassword = os.getenv("MYSQL_REPLICASET_ADMIN_PASSWORD");
const allowedHost = os.getenv("MYSQL_REPLICASET_ALLOWED_HOST");
const target = os.getenv("MYSQL_REPLICASET_TARGET") || "";
const operationTimeout = Number(os.getenv("MYSQL_REPLICASET_TIMEOUT") || "120");
const dryRunOnly = os.getenv("MYSQL_REPLICASET_DRY_RUN_ONLY") === "true";

function connectionOptions(instance) {
  return {
    scheme: "mysql",
    user: adminUser,
    password: adminPassword,
    host: instance.host,
    port: instance.port,
    "ssl-mode": "REQUIRED",
  };
}

function connectTo(instance) {
  shell.connect(connectionOptions(instance));
}

function expectedInstance(name) {
  const match = instances.find((instance) => instance.name === name);
  if (!match) {
    throw new Error(`ReplicaSet target ${name} is not an expected member`);
  }
  return match;
}

function connectToReachableMember() {
  let lastError;
  for (const instance of instances) {
    try {
      connectTo(instance);
      return instance;
    } catch (error) {
      lastError = error;
    }
  }
  throw lastError || new Error("No ReplicaSet member is reachable");
}

function instanceNameForMember(endpoint, member) {
  const address = String(member.address || endpoint).split(":")[0].toLowerCase();
  const label = String(member.label || "").toLowerCase();
  const match = instances.find((instance) => {
    return (
      instance.name.toLowerCase() === label ||
      instance.host.toLowerCase() === address ||
      instance.name.toLowerCase() === address
    );
  });
  return match ? match.name : String(member.label || endpoint);
}

function normalizeStatus(status) {
  const topology = status.replicaSet.topology || {};
  const members = Object.keys(topology).map((endpoint) => {
    const member = topology[endpoint];
    return {
      name: instanceNameForMember(endpoint, member),
      address: member.address || endpoint,
      role: member.instanceRole || "UNKNOWN",
      status: member.status || "UNKNOWN",
      mode: member.mode || null,
      replicationLag: member.replicationLag || null,
      instanceErrors: member.instanceErrors || [],
    };
  });
  const primaryMember = members.find(
    (member) => member.role === "PRIMARY" && member.status === "ONLINE"
  );
  const secondaryMember = members.find(
    (member) => member.role === "SECONDARY" && member.status === "ONLINE"
  );
  return {
    status,
    members,
    primary: primaryMember ? primaryMember.name : null,
    secondary: secondaryMember ? secondaryMember.name : null,
  };
}

function readOnlyState(instance) {
  let session;
  try {
    session = mysql.getSession(connectionOptions(instance));
    const row = session
      .runSql(
        "SELECT @@GLOBAL.read_only, @@GLOBAL.super_read_only, " +
          "@@GLOBAL.gtid_executed, @@GLOBAL.binlog_format, @@GLOBAL.gtid_mode, " +
          "@@GLOBAL.enforce_gtid_consistency, @@GLOBAL.log_replica_updates, " +
          "@@GLOBAL.log_bin, @@GLOBAL.server_id, @@GLOBAL.require_secure_transport"
      )
      .fetchOne();
    return {
      reachable: true,
      readOnly: Number(row[0]),
      superReadOnly: Number(row[1]),
      gtidExecuted: String(row[2]),
      binlogFormat: String(row[3]),
      gtidMode: String(row[4]),
      enforceGtidConsistency: String(row[5]),
      logReplicaUpdates: Number(row[6]),
      logBin: Number(row[7]),
      serverId: Number(row[8]),
      requireSecureTransport: Number(row[9]),
    };
  } catch (error) {
    return {reachable: false, error: String(error.message || error)};
  } finally {
    if (session) {
      session.close();
    }
  }
}

function validateServerConfiguration(instance, state) {
  if (!state.reachable) {
    throw new Error(`${instance.name} is unreachable`);
  }
  if (
    state.gtidMode !== "ON" ||
    state.enforceGtidConsistency !== "ON" ||
    state.binlogFormat !== "ROW" ||
    state.logReplicaUpdates !== 1 ||
    state.logBin !== 1 ||
    state.serverId !== instance.serverId ||
    state.requireSecureTransport !== 1
  ) {
    throw new Error(`${instance.name} does not satisfy the ReplicaSet server policy`);
  }
}

function validateExpectedConfigurations() {
  for (const instance of instances) {
    validateServerConfiguration(instance, readOnlyState(instance));
  }
}

function stateWithServerVariables(rs) {
  const normalized = normalizeStatus(rs.status({extended: 1}));
  normalized.serverVariables = {};
  for (const instance of instances) {
    normalized.serverVariables[instance.name] = readOnlyState(instance);
  }
  return normalized;
}

function validateOnlinePair(normalized) {
  if (normalized.members.length !== instances.length) {
    throw new Error("ReplicaSet does not contain every expected member");
  }
  for (const instance of instances) {
    const matches = normalized.members.filter((member) => member.name === instance.name);
    if (matches.length !== 1) {
      throw new Error(`ReplicaSet membership does not match expected node ${instance.name}`);
    }
  }
  const online = normalized.members.filter((member) => member.status === "ONLINE");
  const primaries = online.filter((member) => member.role === "PRIMARY");
  const secondaries = online.filter((member) => member.role === "SECONDARY");
  if (online.length !== 2 || primaries.length !== 1 || secondaries.length !== 1) {
    throw new Error("ReplicaSet must have one online PRIMARY and one online SECONDARY");
  }
}

function validateWritableTopology(normalized) {
  validateOnlinePair(normalized);
  const primary = normalized.serverVariables[normalized.primary];
  if (!primary.reachable || primary.readOnly !== 0 || primary.superReadOnly !== 0) {
    throw new Error("ReplicaSet PRIMARY is not writable");
  }
  for (const instance of instances) {
    if (instance.name === normalized.primary) {
      continue;
    }
    const secondary = normalized.serverVariables[instance.name];
    if (!secondary.reachable || secondary.readOnly !== 1 || secondary.superReadOnly !== 1) {
      throw new Error("ReplicaSet SECONDARY is not read-only");
    }
  }
}

function validatePostTransition(normalized, expectedPrimary) {
  validateWritableTopology(normalized);
  if (normalized.primary !== expectedPrimary) {
    throw new Error(`Expected ${expectedPrimary} to be PRIMARY after the operation`);
  }
}

function metadataSchemaExists(instance) {
  let session;
  try {
    session = mysql.getSession(connectionOptions(instance));
    const row = session
      .runSql(
        "SELECT COUNT(*) FROM information_schema.schemata " +
          "WHERE schema_name = 'mysql_innodb_cluster_metadata'"
      )
      .fetchOne();
    return Number(row[0]) === 1;
  } finally {
    if (session) {
      session.close();
    }
  }
}

function getReplicaSetIfPresent(instance) {
  if (!metadataSchemaExists(instance)) {
    return null;
  }
  return dba.getReplicaSet();
}

let changed = false;
let rs;
let before = null;

if (action === "check") {
  const connected = connectToReachableMember();
  validateExpectedConfigurations();
  rs = getReplicaSetIfPresent(connected);
  if (!rs) {
    print(
      `PROVISIONING_RESULT=${JSON.stringify({
        action,
        changed: false,
        exists: false,
        status: null,
        members: [],
        primary: null,
        secondary: null,
      })}`
    );
  } else {
    const checked = stateWithServerVariables(rs);
    print(
      `PROVISIONING_RESULT=${JSON.stringify({
        action,
        changed: false,
        exists: true,
        status: checked.status,
        members: checked.members,
        primary: checked.primary,
        secondary: checked.secondary,
        serverVariables: checked.serverVariables,
      })}`
    );
  }
} else {
  if (action === "converge") {
    const seed = expectedInstance(initialPrimary);
    connectTo(seed);
    validateExpectedConfigurations();
    rs = getReplicaSetIfPresent(seed);
    if (!rs) {
      for (const instance of instances) {
        if (instance.name !== seed.name && metadataSchemaExists(instance)) {
          throw new Error(
            `${instance.name} has ReplicaSet metadata while the bootstrap seed does not`
          );
        }
      }
      rs = dba.createReplicaSet(replicaSetName, {
        instanceLabel: initialPrimary,
        replicationAllowedHost: allowedHost,
        replicationSslMode: "REQUIRED",
      });
      changed = true;
    }

    let normalized = normalizeStatus(rs.status({extended: 1}));
    for (const instance of instances) {
      const member = normalized.members.find((candidate) => candidate.name === instance.name);
      if (!member) {
        rs.addInstance(connectionOptions(instance), {
          label: instance.name,
          recoveryMethod: "clone",
        });
        changed = true;
      } else if (member.status === "OFFLINE") {
        rs.rejoinInstance(connectionOptions(instance), {
          recoveryMethod: "incremental",
        });
        changed = true;
      } else if (member.status !== "ONLINE") {
        throw new Error(
          `${instance.name} is ${member.status}; automatic rejoin is not unambiguous`
        );
      }
      normalized = normalizeStatus(rs.status({extended: 1}));
    }
  } else {
    connectToReachableMember();
    if (action !== "failover") {
      validateExpectedConfigurations();
    }
    rs = dba.getReplicaSet();
    before = stateWithServerVariables(rs);
    const targetInstance = expectedInstance(target);

    if (action === "switchover") {
      validateOnlinePair(before);
      if (before.secondary !== target) {
        throw new Error("The planned switchover target must be the current online SECONDARY");
      }
      rs.setPrimaryInstance(connectionOptions(targetInstance), {
        dryRun: true,
        timeout: operationTimeout,
      });
      if (!dryRunOnly) {
        rs.setPrimaryInstance(connectionOptions(targetInstance), {
          timeout: operationTimeout,
        });
        changed = true;
      }
    } else if (action === "failover") {
      if (before.primary === target) {
        throw new Error("The emergency failover target is already PRIMARY");
      }
      if (before.primary) {
        throw new Error(
          `Forced failover is refused while ${before.primary} is still an online PRIMARY`
        );
      }
      const candidate = before.members.find((member) => member.name === target);
      if (!candidate || candidate.status !== "ONLINE" || candidate.role !== "SECONDARY") {
        throw new Error("The emergency failover target must be the current online SECONDARY");
      }
      validateServerConfiguration(targetInstance, before.serverVariables[target]);
      rs.forcePrimaryInstance(connectionOptions(targetInstance), {
        dryRun: true,
        invalidateErrorInstances: true,
        timeout: operationTimeout,
      });
      if (!dryRunOnly) {
        rs.forcePrimaryInstance(connectionOptions(targetInstance), {
          invalidateErrorInstances: true,
          timeout: operationTimeout,
        });
        changed = true;
      }
    } else if (action !== "status") {
      throw new Error(`Unsupported ReplicaSet action ${action}`);
    }
  }

  const current = dryRunOnly && before ? before : stateWithServerVariables(rs);
  if (action === "converge" || action === "status") {
    validateWritableTopology(current);
  } else if (action === "switchover" && !dryRunOnly) {
    validatePostTransition(current, target);
  } else if (action === "failover" && !dryRunOnly) {
    if (current.primary !== target) {
      throw new Error(`Expected ${target} to be PRIMARY after forced failover`);
    }
    const newPrimary = current.serverVariables[target];
    validateServerConfiguration(targetInstance, newPrimary);
    if (!newPrimary.reachable || newPrimary.readOnly !== 0 || newPrimary.superReadOnly !== 0) {
      throw new Error("Forced PRIMARY is not writable");
    }
  }

  print(
    `PROVISIONING_RESULT=${JSON.stringify({
      action,
      changed,
      dryRunOnly,
      exists: true,
      before,
      status: current.status,
      members: current.members,
      primary: current.primary,
      secondary: current.secondary,
      serverVariables: current.serverVariables,
    })}`
  );
}

File: roles/components/mysql_replicaset/files/mysql-replicaset.py
"""Manage an InnoDB ReplicaSet through MySQL Shell's Python AdminAPI."""

import json
import os


def normalize_status(status, instances):
    members = []
    for endpoint, member in status['replicaSet'].get('topology', {}).items():
        address = str(member.get('address') or endpoint).split(':')[0].lower()
        label = str(member.get('label') or '').lower()
        instance = next((item for item in instances if (
            item['name'].lower() == label
            or item['host'].lower() == address
            or item['name'].lower() == address
        )), None)
        members.append({
            'name': instance['name'] if instance else str(member.get('label') or endpoint),
            'address': member.get('address') or endpoint,
            'role': member.get('instanceRole') or 'UNKNOWN',
            'status': member.get('status') or 'UNKNOWN',
            'mode': member.get('mode'),
            'replicationLag': member.get('replicationLag'),
            'instanceErrors': member.get('instanceErrors') or [],
        })
    return {
        'status': status,
        'members': members,
        'primary': next((m['name'] for m in members if m['role'] == 'PRIMARY' and m['status'] == 'ONLINE'), None),
        'secondary': next((m['name'] for m in members if m['role'] == 'SECONDARY' and m['status'] == 'ONLINE'), None),
    }


def validate_server_configuration(instance, state):
    if not state['reachable']:
        raise RuntimeError(f"{instance['name']} is unreachable")
    expected = {'gtidMode': 'ON', 'enforceGtidConsistency': 'ON', 'binlogFormat': 'ROW',
                'logReplicaUpdates': 1, 'logBin': 1, 'serverId': instance['serverId'],
                'requireSecureTransport': 1}
    if any(state[key] != value for key, value in expected.items()):
        raise RuntimeError(f"{instance['name']} does not satisfy the ReplicaSet server policy")


def validate_online_pair(state, instances):
    members = state['members']
    if len(members) != len(instances):
        raise RuntimeError('ReplicaSet does not contain every expected member')
    for instance in instances:
        if sum(m['name'] == instance['name'] for m in members) != 1:
            raise RuntimeError(f"ReplicaSet membership does not match expected node {instance['name']}")
    online = [m for m in members if m['status'] == 'ONLINE']
    if (len(online) != 2 or sum(m['role'] == 'PRIMARY' for m in online) != 1
            or sum(m['role'] == 'SECONDARY' for m in online) != 1):
        raise RuntimeError('ReplicaSet must have one online PRIMARY and one online SECONDARY')


def validate_writable_topology(state, instances):
    validate_online_pair(state, instances)
    for instance in instances:
        variables = state['serverVariables'][instance['name']]
        expected = 0 if instance['name'] == state['primary'] else 1
        if (not variables['reachable'] or variables['readOnly'] != expected
                or variables['superReadOnly'] != expected):
            raise RuntimeError(f"{instance['name']} has incorrect write protection")


def validate_failover_candidate(state, target):
    if state['primary']:
        raise RuntimeError('Forced failover is refused while an online PRIMARY exists')
    candidate = next((m for m in state['members'] if m['name'] == target), None)
    # A replica reconnecting to the unavailable PRIMARY is CONNECTING.
    # AdminAPI dry-run checks its transaction set and applier before promotion.
    if (not candidate or candidate['status'] not in ('ONLINE', 'CONNECTING')
            or candidate['role'] != 'SECONDARY'):
        raise RuntimeError('The emergency failover target must be a reachable SECONDARY')


class ReplicaSetManager:
    def __init__(self, shell, dba, mysql, environment):
        self.shell, self.dba, self.mysql = shell, dba, mysql
        self.instances = json.loads(environment.get('MYSQL_REPLICASET_INSTANCES', '[]'))
        self.name = environment['MYSQL_REPLICASET_NAME']
        self.initial_primary = environment['MYSQL_REPLICASET_INITIAL_PRIMARY']
        self.admin_user = environment['MYSQL_REPLICASET_ADMIN_USER']
        self.admin_password = environment['MYSQL_REPLICASET_ADMIN_PASSWORD']
        self.allowed_host = environment['MYSQL_REPLICASET_ALLOWED_HOST']
        self.target = environment.get('MYSQL_REPLICASET_TARGET', '')
        self.timeout = int(environment.get('MYSQL_REPLICASET_TIMEOUT', '120'))
        self.dry_run = environment.get('MYSQL_REPLICASET_DRY_RUN_ONLY') == 'true'

    def expected_instance(self, name):
        for instance in self.instances:
            if instance['name'] == name:
                return instance
        raise RuntimeError(f'ReplicaSet target {name} is not an expected member')

    def connection_options(self, instance):
        return {'scheme': 'mysql', 'user': self.admin_user, 'password': self.admin_password,
                'host': instance['host'], 'port': instance['port'], 'ssl-mode': 'REQUIRED'}

    def connect_to(self, instance):
        self.shell.connect(self.connection_options(instance))

    def connect_to_reachable_member(self):
        for instance in self.instances:
            try:
                self.connect_to(instance)
                return instance
            except Exception:
                continue
        raise RuntimeError('No ReplicaSet member is reachable')

    def read_only_state(self, instance):
        session = None
        try:
            session = self.mysql.get_session(self.connection_options(instance))
            row = session.run_sql(
                'SELECT @@GLOBAL.read_only, @@GLOBAL.super_read_only, '
                '@@GLOBAL.gtid_executed, @@GLOBAL.binlog_format, @@GLOBAL.gtid_mode, '
                '@@GLOBAL.enforce_gtid_consistency, @@GLOBAL.log_replica_updates, '
                '@@GLOBAL.log_bin, @@GLOBAL.server_id, @@GLOBAL.require_secure_transport'
            ).fetch_one()
            return {'reachable': True, 'readOnly': int(row[0]), 'superReadOnly': int(row[1]),
                    'gtidExecuted': str(row[2]), 'binlogFormat': str(row[3]), 'gtidMode': str(row[4]),
                    'enforceGtidConsistency': str(row[5]), 'logReplicaUpdates': int(row[6]),
                    'logBin': int(row[7]), 'serverId': int(row[8]), 'requireSecureTransport': int(row[9])}
        except Exception:
            return {'reachable': False}
        finally:
            if session is not None:
                session.close()

    def validate_expected_configurations(self):
        for instance in self.instances:
            validate_server_configuration(instance, self.read_only_state(instance))

    def state(self, replicaset):
        state = normalize_status(replicaset.status({'extended': 1}), self.instances)
        state['serverVariables'] = {i['name']: self.read_only_state(i) for i in self.instances}
        return state

    def metadata_schema_exists(self, instance):
        session = self.mysql.get_session(self.connection_options(instance))
        try:
            row = session.run_sql(
                "SELECT COUNT(*) FROM information_schema.schemata "
                "WHERE schema_name = 'mysql_innodb_cluster_metadata'"
            ).fetch_one()
            return int(row[0]) == 1
        finally:
            session.close()

    def get_replicaset_if_present(self, instance):
        return self.dba.get_replica_set() if self.metadata_schema_exists(instance) else None

    def run(self, action):
        self.shell.options.useWizards = False
        changed, before = False, None
        if action == 'check':
            connected = self.connect_to_reachable_member()
            self.validate_expected_configurations()
            rs = self.get_replicaset_if_present(connected)
            if rs is None:
                return {'action': action, 'changed': False, 'exists': False, 'status': None,
                        'members': [], 'primary': None, 'secondary': None}
            return {'action': action, 'changed': False, 'exists': True, **self.state(rs)}
        if action == 'converge':
            seed = self.expected_instance(self.initial_primary)
            self.connect_to(seed)
            self.validate_expected_configurations()
            rs = self.get_replicaset_if_present(seed)
            if rs is None:
                for instance in self.instances:
                    if instance['name'] != seed['name'] and self.metadata_schema_exists(instance):
                        raise RuntimeError(f"{instance['name']} has ReplicaSet metadata while the bootstrap seed does not")
                rs = self.dba.create_replica_set(self.name, {
                    'instanceLabel': self.initial_primary, 'replicationAllowedHost': self.allowed_host,
                    'replicationSslMode': 'REQUIRED',
                })
                changed = True
            state = normalize_status(rs.status({'extended': 1}), self.instances)
            for instance in self.instances:
                member = next((m for m in state['members'] if m['name'] == instance['name']), None)
                address = f"{instance['host']}:{instance['port']}"
                if member is None:
                    rs.add_instance(address, {'label': instance['name'], 'recoveryMethod': 'clone'})
                    changed = True
                elif member['status'] == 'OFFLINE':
                    rs.rejoin_instance(address, {'recoveryMethod': 'incremental'})
                    changed = True
                elif member['status'] != 'ONLINE':
                    raise RuntimeError(f"{instance['name']} is {member['status']}; automatic rejoin is not unambiguous")
                state = normalize_status(rs.status({'extended': 1}), self.instances)
        elif action in ('status', 'switchover', 'failover'):
            self.connect_to_reachable_member()
            if action != 'failover':
                self.validate_expected_configurations()
            rs = self.dba.get_replica_set()
            before = self.state(rs)
            if action in ('switchover', 'failover'):
                target = self.expected_instance(self.target)
                address = f"{target['host']}:{target['port']}"
                options = {'timeout': self.timeout}
                if action == 'switchover':
                    validate_online_pair(before, self.instances)
                    if before['secondary'] != self.target:
                        raise RuntimeError('The planned switchover target must be the current online SECONDARY')
                    transition = rs.set_primary_instance
                else:
                    validate_failover_candidate(before, self.target)
                    validate_server_configuration(target, before['serverVariables'][self.target])
                    transition = rs.force_primary_instance
                    options['invalidateErrorInstances'] = True
                transition(address, {**options, 'dryRun': True})
                if not self.dry_run:
                    transition(address, options)
                    changed = True
        else:
            raise RuntimeError(f'Unsupported ReplicaSet action {action}')

        current = before if self.dry_run and before is not None else self.state(rs)
        if action in ('converge', 'status') or (action == 'switchover' and not self.dry_run):
            validate_writable_topology(current, self.instances)
        if action in ('switchover', 'failover') and not self.dry_run:
            if current['primary'] != self.target:
                raise RuntimeError(f'Expected {self.target} to be PRIMARY after the operation')
            if action == 'failover':
                primary = current['serverVariables'][self.target]
                validate_server_configuration(self.expected_instance(self.target), primary)
                if primary['readOnly'] != 0 or primary['superReadOnly'] != 0:
                    raise RuntimeError('Forced PRIMARY is not writable')
        return {'action': action, 'changed': changed, 'dryRunOnly': self.dry_run,
                'exists': True, 'before': before, **current}


def main():
    from mysqlsh import globals as shell_globals, mysql

    manager = ReplicaSetManager(shell_globals.shell, shell_globals.dba, mysql, os.environ)
    result = manager.run(os.environ.get('MYSQL_REPLICASET_ACTION', 'status'))
    print('PROVISIONING_RESULT=' + json.dumps(result, separators=(',', ':')))


if __name__ == '__main__':
    main()
